### PR TITLE
Hadoop InputRowParser for Orc file

### PR DESCRIPTION
**Pull request by sirpkt**
Related with #3017 

- user SHOULD supply schema information as a form of string representation of Orc struct type info.
  for example, string column `col1` and array of string column `col2` is represented by `struct<col1:string,col2:array<string>>`

- only support java primitive type columns and array of java primitive type columns

- as shown in `hadoop_orc_job.json` example, inputFormat Should be set as `org.apache.hadoop.hive.ql.io.orc.OrcNewInputFormat`

### CODE REVIEW

**Comment by fjy** [comment]
@sirpkt can we also get some docs on usage and also update extensions.md file?

**Comment by sirpkt** [comment]
@fjy Sure, I'll update doc.

**Comment by sirpkt** [comment]
This patch used hive orc reader so that it has complex dependencies including hive-exec.
Apache Orc(https://github.com/apache/orc) is working to separate Orc code from Hive but only done for org.apache.hadoop.mapred (old API) currently.
As pull request for org.apache.hadoop.mapreduce already exists, I think it will soon support new MapReduce API.
I'll watch the progress and update the patch to use that library.

**Comment by sirpkt** [comment]
During testing in the real server cluster, I experienced library dependency problem.
So, I close this PR until I resolve that problem.

**Comment by sirpkt** [comment]
maven dependency updated

File: docs/content/development/extensions-contrib/orc.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+```json
+{
+  "type": "index_hadoop",
+  "spec": {
+    "ioConfig": {
+      "type": "hadoop",
+      "inputSpec": {
+        "type": "static",
+        "inputFormat": "org.apache.hadoop.hive.ql.io.orc.OrcNewInputFormat",
+        "paths": "no_metrics"
```
Comment (fjy) [inline docs/content/development/extensions-contrib/orc.md]: what should be filled out in the paths field?

File: docs/content/development/extensions.md [hunk 2]
```diff
@@ -51,6 +51,7 @@ All of these community extensions can be downloaded using *pull-deps* with the c
 |druid-rabbitmq|RabbitMQ firehose.|[link](../development/extensions-contrib/rabbitmq.html)|
 |druid-rocketmq|RocketMQ firehose.|[link](../development/extensions-contrib/rocketmq.html)|
 |graphite-emitter|Graphite metrics emitter|[link](../development/extensions-contrib/graphite.html)|
+|druid-orc-extensions|Support for data in Apache Orc data format.|[link](../development/extensions-contrib/orc.html)|
```
Comment (fjy) [inline docs/content/development/extensions.md]: small nit, but can we make the list lexicographic?

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    TimestampSpec timestampSpec = parseSpec.getTimestampSpec();
+    DateTime dateTime = timestampSpec.extractTimestamp(map);
+
+    return new MapBasedInputRow(dateTime, dimensions, map);
+  }
+
+  private void initialize()
+  {
+    if (typeString == null)
```
Comment (fjy) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: there's a formatting problem here

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    initialize();
+  }
+
+  @Override
+  public InputRow parse(OrcStruct input)
+  {
+    Map<String, Object> map = Maps.newHashMap();
+    List<? extends StructField> fields = oip.getAllStructFieldRefs();
+    for (StructField field: fields)
+    {
```
Comment (fjy) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: formatting problem here and in other places

File: extensions-contrib/orc-extensions/src/test/java/io/druid/data/input/orc/OrcIndexGeneratorJobTest.java [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+import io.druid.segment.QueryableIndexIndexableAdapter;
+import io.druid.segment.Rowboat;
+import io.druid.segment.indexing.DataSchema;
+import io.druid.segment.indexing.granularity.UniformGranularitySpec;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.partition.HashBasedNumberedShardSpec;
+import io.druid.timeline.partition.ShardSpec;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.ql.exec.vector.*;
```
Comment (fjy) [inline extensions-contrib/orc-extensions/src/test/java/io/druid/data/input/orc/OrcIndexGeneratorJobTest.java]: we avoid including *s, please remove

File: pom.xml [hunk 6]
```diff
@@ -104,6 +104,7 @@
         <module>extensions-contrib/distinctcount</module>
         <module>extensions-contrib/parquet-extensions</module>
 	<module>extensions-contrib/statsd-emitter</module>
+        <module>extensions-contrib/orc-extensions</module>
```
Comment (fjy) [inline pom.xml]: can we keep these in lexicographic order?

**Comment by fjy** [comment]
overall this PR looks reasonable to me however I have little experience with ORC files. Can someone who has more experience with ORC files take a look and comment on the API?

**Comment by fjy** [comment]
@sirpkt there are some merge errors

**Comment by sirpkt** [comment]
rebase and update based on comments

File: extensions-contrib/orc-extensions/src/test/java/io/druid/data/input/orc/OrcIndexGeneratorJobTest.java (hunk 5 above)
Comment (fjy) [inline extensions-contrib/orc-extensions/src/test/java/io/druid/data/input/orc/OrcIndexGeneratorJobTest.java]: @sirpkt

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    return new OrcHadoopInputRowParser(parseSpec, typeString);
+  }
+
+  public static String typeStringFromParseSpec(ParseSpec parseSpec)
+  {
+    StringBuilder builder = new StringBuilder("struct<");
+    builder.append(parseSpec.getTimestampSpec()).append(":string");
+    if (parseSpec.getDimensionsSpec().getDimensionNames().size() > 0) {
+      builder.append(",");
+      builder.append(StringUtils.join(parseSpec.getDimensionsSpec().getDimensionNames(), ":string,")).append(":string");
```
Comment (nishantmonu51) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: do we need to append ":string" twice here ?

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+  {
+    return new OrcHadoopInputRowParser(parseSpec, typeString);
+  }
+
+  public InputRowParser withTypeString(String typeString)
+  {
+    return new OrcHadoopInputRowParser(parseSpec, typeString);
+  }
+
+  public static String typeStringFromParseSpec(ParseSpec parseSpec)
```
Comment (nishantmonu51) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: please add UT for this method

File: extensions-contrib/orc-extensions/pom.xml [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        <dependency>
+            <groupId>io.druid</groupId>
+            <artifactId>druid-indexing-hadoop</artifactId>
+            <version>${project.parent.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-exec</artifactId>
+            <version>2.0.0</version>
```
Comment (nishantmonu51) [inline extensions-contrib/orc-extensions/pom.xml]: minor nit : would be nice to specify the version as a property in pom file - 
 <hive.compile.version>2.0.0</hive.compile.version>

File: docs/content/development/extensions-contrib/orc.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+          },
+          "dimensionsSpec": {
+            "dimensions": [
+              "name"
+            ],
+            "dimensionExclusions": [],
+            "spatialDimensions": []
+          }
+        },
+        "typeString": "struct<time:string,name:string>"
```
Comment (nishantmonu51) [inline docs/content/development/extensions-contrib/orc.md]: I wonder if we can generate this String using dimensionsSpec, it includes the DimensionSchema that has the type info also ?

**Comment by fjy** [comment]
@sirpkt any chance we can finish this up?

**Comment by fjy** [comment]
@sirpkt

**Comment by sirpkt** [comment]
@fjy sorry for late response.
I just moved my house so had not enough time to follow review.
I'll update based on the review comment

File: docs/content/development/extensions-contrib/orc.md (hunk 10 above)
Comment (sirpkt) [inline docs/content/development/extensions-contrib/orc.md]: @nishantmonu51 String dimension of Druid can have either single String value or list of String values.
In Orc, however, those two have different structure.
So, I don't think it is possible to generate type string from dimensionsSpec.

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java (hunk 8 above)
Comment (sirpkt) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: @nishantmonu51 I added UT and found bug in the code! (and fixed it)
Thanks.

**Comment by sirpkt** [comment]
updated based on the review comments

File: docs/content/development/extensions-contrib/orc.md (hunk 10 above)
Comment (nishantmonu51) [inline docs/content/development/extensions-contrib/orc.md]: ah, ok.

**Comment by nishantmonu51** [comment]
👍 , all my comments are addressed.

File: extensions-contrib/orc-extensions/pom.xml [hunk 11]
```diff
@@ -0,0 +1,134 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fjy) [inline extensions-contrib/orc-extensions/pom.xml]: this is missing a license header

**Comment by fjy** [comment]
👍  after the missing license header is added

**Comment by fjy** [comment]
@sirpkt

**Comment by sirpkt** [comment]
@fjy sorry for late response.
I rebased code and added license header.

**Comment by b-slim** [comment]
@sirpkt can you help with this issue https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/druid-user/wiXVMMmXOgU/fHXn6hR0CQAJ

**Pull request merged**